### PR TITLE
Use Vector256 rather than variable length Vector

### PR DESCRIPTION
**Pull request by benaadams**
## Changes

- Use Vector256 rather than Vector and enabling based on whether length is 32

#### What types of changes does your code introduce?

- [ ] Bugfix (a non-breaking change that fixes an issue)
- [ ] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [x] Optimization
- [ ] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [ ] Other: _Description_

#### Requires testing

- [ ] Yes
- [x] No

#### If yes, did you write tests?

- [ ] Yes
- [x] No

### CODE REVIEW

File: src/Nethermind/Nethermind.Evm/EvmStack.cs [hunk 1]
```diff
@@ -301,6 +321,17 @@ public Address PopAddress()
...
 
+        public Vector256<byte> PopVector256()
+        {
+            if (Head-- == 0)
+            {
+                Metrics.EvmExceptions++;
+                ThrowEvmStackUnderflowException();
+            }
+
+            return Unsafe.ReadUnaligned<Vector256<byte>>(ref Unsafe.Add(ref MemoryMarshal.GetReference(_bytes), Head * 32));
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Evm/EvmStack.cs:332]: Question: Before we returned Span<byte> which was 12 bytes, now we are returning 32 bytes. All on stack. Doesn't it lead to more data moving to stack?

File: src/Nethermind/Nethermind.Evm/VirtualMachine.cs [hunk 2]
```diff
@@ -881,7 +876,7 @@ static void UpdateCurrentState(EvmState state, int pc, long gas, int stackHead)
                             BytesMax32.AsSpan(0, position).CopyTo(b[..position]);
                         }
 
-                        stack.PushBytes(b);
+                        stack.PushWord256(b);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Evm/VirtualMachine.cs:879]: Can above if-esle be done with Vector256? I think so, we need to copy part and 0 other bytes?

File: src/Nethermind/Nethermind.Evm/VirtualMachine.cs [hunk 3]
```diff
@@ -990,114 +985,48 @@ static void UpdateCurrentState(EvmState state, int pc, long gas, int stackHead)
...
+                        Vector256<byte> bVec = stack.PopVector256();
 
-                            Vector.BitwiseAnd(aVec, bVec).CopyTo(stack.Register);
-                        }
-                        else
-                        {
-                            ref ulong refA = ref MemoryMarshal.AsRef<ulong>(a);
-                            ref ulong refB = ref MemoryMarshal.AsRef<ulong>(b);
-                            ref ulong refBuffer = ref MemoryMarshal.AsRef<ulong>(stack.Register);
+                        Unsafe.WriteUnaligned(ref MemoryMarshal.GetReference(stack.Register), Vector256.BitwiseAnd(aVec, bVec));
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Evm/VirtualMachine.cs:991]: Maybe Stack.Register should be Vector256? Probably yes.

File: src/Nethermind/Nethermind.Evm/VirtualMachine.cs [hunk 4]
```diff
@@ -1135,7 +1064,7 @@ static void UpdateCurrentState(EvmState state, int pc, long gas, int stackHead)
                         if (!UpdateMemoryCost(vmState, ref gasAvailable, in memSrc, memLength)) goto OutOfGas;
 
                         Span<byte> memData = vmState.Memory.LoadSpan(in memSrc, memLength);
-                        stack.PushBytes(ValueKeccak.Compute(memData).BytesAsSpan);
+                        stack.PushWord256(ValueKeccak.Compute(memData).BytesAsSpan);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Evm/VirtualMachine.cs:1067]: Why not push it as Vector256, skipping those conversions?

File: src/Nethermind/Nethermind.Evm/EvmStack.cs [hunk 5]
```diff
@@ -321,10 +321,15 @@ public Address PopAddress()
             return ref _bytes[Head * 32];
         }
 
-        [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public Vector256<byte> PopVector256()
         {
-            return Unsafe.ReadUnaligned<Vector256<byte>>(ref MemoryMarshal.GetReference(PopWord256()));
+            if (Head-- == 0)
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Evm/EvmStack.cs:326]: should we decrement the Head if it is already 0?

**Comment by LukaszRozmej** [comment]
Let's run hive tests before merging.

**Comment by benaadams** [comment]
Replacing with https://github.com/NethermindEth/nethermind/pull/5826

**Pull request closed without merging**